website_portal_contract/controllers/main.py

```diff
@@ -42,29 +40,19 @@ def portal_my_contracts(self):
         )
 
     @http.route(
-        ['/contract/<int:id>'],
+        ['/contract/<model("account.analytic.account"):contract>'],
         type='http',
         auth='user',
         website=True
     )
-    def portal_contract(self, pdf=None, **kwargs):
-        rec_id = kwargs.get('id')
-
-        if not rec_id:
-            return json.dumps(False)
-
-        contract = request.env['account.analytic.account'].browse(rec_id)
-
-        if not contract:
-            return request.render('website.404')
-
+    def portal_contract(self, contract):
         action = request.env.ref(
             'contract.action_account_analytic_overdue_all'
-        ).id
+        )
         values = {
             'user': request.env.user,
             'contract': contract,
-            'action': action,
+            'action': action.id,
             'invoices': contract.recurring_invoice_line_ids.ids,
```

Comment:
Why `.ids`? You should be using recordsets basically everywhere, or at least add `_ids` to the end of your var